### ISSUE TITLE
Missing support for signing NuGet packages

### ISSUE THREAD

**Issue by MichaelGrafnetter** [Azure/trusted-signing-action#53]
### Current Behavior

When trying to sign `*.nupkg` and `*.snupkg` packages, I get the following error:

```txt
SignTool Error: This file format cannot be signed because it is not recognized.
```

### Expected Behavior

It would be great to be able to sign .NET DLLs and packages using the same tool.

**Comment by japarson** [comment]
Hi @MichaelGrafnetter,

Unfortunately, signing NuGet packages with Trusted Signing is difficult right now because NuGet requires a thumbprint certificate and Trusted Signing certificates are short-lived, requiring manual updates every 3 days.

There is an open issue to improve this system here: https://github.com/NuGet/Home/pull/13791

**Comment by MichaelGrafnetter** [comment]
Hi @japarson , thanks for the response. Yes, I am manually registering the certificates in NuGet Gallery with each release and it is extremely cumbersome.

For now, I am successfully signing NuGet packages using the `sign code trusted-signing ...` command executed from GitHub Actions. I wanted to switch to the `Azure/trusted-signing-action`, but it does not seem to support NuGet packages at all.

**Comment by FabianBoumeester** [comment]
Hi @MichaelGrafnetter , @japarson ,

Has there been any movement on this issue? Pushing a package to nuget, nowadays requires signed packages. So this is one of the featues I’m looking for mostly. Is SignTool doesn’t support nuget packages is there another signing tool that can be leveraged by TrustdSigning? 

I managed to configure my azure pipeline to sign just about everything, except the nuget packages.

-Fabian

**Comment by Jaxelr** [comment]
There has been progress, but we still need support from the nuget team to get this done. A [proposal  has been accepted](https://github.com/NuGet/Home/blob/dev/accepted/2024/improve-support-for-packages-signed-with-trusted-signing.md) which outlines the planned improvements for supporting trusted signing.

**Comment by LanceMcCarthy** [comment]
@Jaxelr How far along are you? Using the .NET sign CLI is not possible because it requires HSM-backed Azure KeyVault (suuuper expensive, I learned a $4000 mistake once enabling HSM)

**Comment by Jaxelr** [comment]
How come? Sign Cli supports [artifact signing](https://github.com/dotnet/sign/blob/main/docs/artifact-signing-integration.md).

**Comment by LanceMcCarthy** [comment]
@Jaxelr with `sign` You need a premium Az Storage account because it requires HSM-backed keys. These accounts are minimum $2500 a month (I learned the hard way, ended up with a $4000 bill before I noticed).

**Comment by Jaxelr** [comment]
No, you can use your artifact signing account only to perform the signing, no need to buy anything else. Case and point, if you sign a vsix using this github action, underneath we use the sign cli.

**Comment by LanceMcCarthy** [comment]
@Jaxelr Please show me an example where it works for nupkg

Edit
I am mixing up Azure Trusted Signing (ATS) with Azure Artifact Signing. ATS is the one that doesn't support nupkg and what I was referring to earlier. I will try creating a new cert in AAS and see if it works. Thank you!

--If you're referring to msix, vsix, etc, yeah, thats fine. They all work, I've been using it since early private testing under MVP NDA, ect. The problem with this open issue is that it does not work for nukg because you need to have a link to nuget.org to register the short-lived cert's thumbprint with them. I tried using my signing account for nupkg and it fails. I use the same account for msix nearly every day, and it's flawless (absolutely love it). One thought I had was to zip up the nupkg, sign the zip, unzip it and then submit it. but, we run into the same problem... nuget.org won't accept it because the thumbprint of the cert is not listed in my nuget.org account.--

**Comment by bgrainger** [comment]
@LanceMcCarthy Until https://github.com/NuGet/NuGetGallery/issues/10027 is fixed, I think the only workaround is to:
* unzip the nupkg
* sign each DLL individually with `sign` using Azure Artifact Signing (aka Trusted Signing)
* zip the nupkg back up with the signed DLLs
* (don't sign the nupkg itself)

I have a pwsh script that basically does:

```pwsh
	Expand-Archive -Path $package -DestinationPath $packagingRoot
	foreach ($dllPath in Get-ChildItem -Path $packagingRoot -Filter "*.dll" -Recurse) {
		# dotnet sign
	}
	Compress-Archive -Path "$packagingRoot\*" -DestinationPath $repackagedPath -Force
```

**Comment by Jaxelr** [comment]
Worth pointing out that you could use [wildcards](https://github.com/dotnet/sign/blob/main/docs/file-globbing.md) and max concurrency with sign cli.
